src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.PieceDirectory.cs

```diff
@@ -0,0 +1,456 @@
+        //  At init time, the first piece descriptor is loaded into _pieceStreamInfoArray. Thereafter, descriptors are loaded
+        //  on demand in piece order and without any gap between them.
+
+        private sealed class PieceDirectory
+        {
+            /// <summary>
+            /// PieceStreamInfo Descriptor.
+            /// Provides access to piece's stream and start offset.
+            /// </summary>
+            /// <remarks>
+            /// This is a private class. All of the above information, together with position
+            /// information (piece number, last-piece status), is accessed through the
+            /// PieceDirectory class by providing offsets or piece numbers.
+            /// </remarks>
+            private sealed class PieceStreamInfo : IComparable<PieceStreamInfo>
+            {
+
+                /// <summary>
+                /// Initialize a PieceStreamInfo.
+                /// </summary>
+                internal PieceStreamInfo(Stream stream, long pieceStart)
+                {
+                    Debug.Assert(stream != null);
+                    Debug.Assert(pieceStart >= 0);
+                    Stream = stream;
+                    StartOffset = pieceStart;
+                }
+
+                /// <summary>
+                /// Returns the startOffset of the piece in the part
+                /// </summary>
+                internal long StartOffset { get; }
+
+
+                /// <summary>
+                /// Returns the stream for this piece
+                /// </summary>
+                internal Stream Stream { get; set; }
+
+
+                /// <summary>
+                /// IComparable.CompareTo implementation which allows sorting
+                /// descriptors by range of offsets.
+                /// </summary>
+                int IComparable<PieceStreamInfo>.CompareTo(PieceStreamInfo? pieceStreamInfo)
+                {
+                    return pieceStreamInfo == null
+                        ? 1
+                        : StartOffset.CompareTo(pieceStreamInfo.StartOffset);
+                }
+
+            }
+
+            /// <summary>
+            /// Load the descriptor of the first piece into _pieceStreamInfoArray and store context data
+            /// extracted from the arguments.
+            /// </summary>
+            internal PieceDirectory(List<ZipPackagePartPiece> sortedPieceInfoList, ZipStreamManager zipStreamManager, FileAccess access)
+            {
+                if (sortedPieceInfoList == null)
+                    throw new ArgumentNullException(nameof(sortedPieceInfoList));
+
+                Debug.Assert(sortedPieceInfoList.Count > 0);
+
+                // Initialize the first piece.
+                _pieceStreamInfoList = [
+                    new PieceStreamInfo(
+                    zipStreamManager.Open(sortedPieceInfoList[0].ZipArchiveEntry, access),
+                    0 /*startOffset*/)
+                ];
```

Comment:
Style: `[` on new line, indent for parameters of the constructor call, use `startOffset: 0` for naming parameter